base/iterators.jl

```diff
@@ -1095,4 +1094,35 @@ eltype(::Type{Stateful{T, VS}} where VS) where {T} = eltype(T)
 IteratorEltype(::Type{Stateful{T,VS}}) where {T,VS} = IteratorEltype(T)
 length(s::Stateful) = length(s.itr) - s.taken
 
+"""
+    only(x)
+
+Returns the one and only element of collection `x`, and throws an `ArgumentError` if the
+collection has zero or multiple elements.
+"""
+Base.@propagate_inbounds function only(x)
+    i = iterate(x)
+    @boundscheck if i === nothing
+        throw(ArgumentError("Collection is empty, must contain exactly 1 element"))
+    end
+    (ret, state) = i
+    j = iterate(x, state)
+    @boundscheck if j !== nothing
+        throw(ArgumentError("Collection has multiple elements, must contain exactly 1 element"))
+    end
+    return ret
+end
+
+# Collections of known size
```

Comment:
```suggestion
# Collections of known size
only(x::Ref) = x[]
```

Reply:
good catch! (assuming I am checking this right)
```julia
julia> @code_typed only(1)
CodeInfo(
1 ─     goto #4 if not $(Expr(:boundscheck))
2 ─     goto #4 if not false
3 ─     nothing::Nothing
4 ┄     goto #7 if not $(Expr(:boundscheck))
5 ─     goto #7 if not false
6 ─     nothing::Nothing
7 ┄     return x
) => Int64

julia> only(x::Number) = x
only (generic function with 8 methods)

julia> @code_typed only(1)
CodeInfo(
1 ─     return x
) => Int64

julia> @code_typed only(Ref([1, 2]))
CodeInfo(
1 ─ %1 = Base.getfield(x, :x)::Array{Int64,1}
└──      goto #4 if not $(Expr(:boundscheck))
2 ─      goto #4 if not false
3 ─      nothing::Nothing
4 ┄      goto #7 if not $(Expr(:boundscheck))
5 ─      goto #7 if not false
6 ─      nothing::Nothing
7 ┄      return %1
) => Array{Int64,1}

julia> only(x::Ref) = x[]
only (generic function with 9 methods)

julia> @code_typed only(Ref([1, 2]))
CodeInfo(
1 ─ %1 = Base.getfield(x, :x)::Array{Int64,1}
└──      return %1
) => Array{Int64,1}
```

Reply:
I gues we should also add `only(x::Char) = x`

given
```julia
julia> only('a')
'a': ASCII/Unicode U+0061 (category Ll: Letter, lowercase)

julia> @code_typed only('a')
CodeInfo(
1 ──       goto #4 if not $(Expr(:boundscheck))
2 ──       goto #4 if not false
3 ──       nothing::Nothing
4 ┄─       goto #10 if not $(Expr(:boundscheck))
5 ──       goto #7 if not true
6 ──       goto #8
7 ──       goto #8
8 ┄─ %8  = φ (#6 => true, #7 => false)::Bool
│    %9  = Core.Intrinsics.not_int(%8)::Bool
└───       goto #10 if not %9
9 ── %11 = %new(Core.ArgumentError, "Collection has multiple elements, must contain exactly 1 element")::ArgumentError
│          Main.throw(%11)::Union{}
└───       $(Expr(:unreachable))::Union{}
10 ┄       return x
) => Char

julia> only(x::Char) = x
only (generic function with 10 methods)

julia> @code_typed only('a')
CodeInfo(
1 ─     return x
) => Char
```

Reply:
added - thanks!